### PR TITLE
feat(text editor mentions): add plugin factory

### PR DESCRIPTION
**Pull request by john-traas**
Add's a new custom node spec to use for mentions in the Text Editor.

Fixes https://github.com/Lundalogik/crm-feature/issues/4311

### CODE REVIEW

File: src/components/text-editor/utils/markdown-converter.ts [hunk 1]
```diff
@@ -1,14 +1,25 @@
...
     MarkdownSerializer,
+    MarkdownSerializerState,
     defaultMarkdownSerializer,
 } from 'prosemirror-markdown';
 import { markdownToHTML } from '../../markdown/markdown-parser';
 
+const mentionSerializer = {
+    mention: (state: MarkdownSerializerState, node: ProseMirrorNode) => {
+        state.write(
+            `<limebb-mention type="${node.attrs.type}" objectid="${node.attrs.objectid}" descriptive="${node.attrs.descriptive}"></limebb-mention>`,
```
Comment (jgroth) [inline src/components/text-editor/utils/markdown-converter.ts:14]: I don't think we should hardcode this element here like this. For this to be reusable, we would either have to provide some config or a factory to the the text editor component that will create the element that the consumer needs

Comment (john-traas) [reply src/components/text-editor/utils/markdown-converter.ts:14]: Absolutely

File: src/components/text-editor/types.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { NodeSpec } from 'prosemirror-model';
+
+/**
+ * @private
+ * set to private to avoid usage while under development
+ */
+export type TextEditorPlugin = {
+    node: NodeConfig[];
```
Comment (jgroth) [inline src/components/text-editor/types.ts:8]: We can't expose types from a third party like this. If there's a breaking change in these types those would be immediately exposed from lime elements without us necessarily being aware of it. We need to create our own type and then map that type to the `NodeSpec` type

File: src/components/text-editor/text-editor.tsx [hunk 3]
```diff
@@ -93,6 +95,9 @@ export class TextEditor implements FormComponent<string> {
     @Prop({ reflect: true })
     public value: string;
 
+    @Prop()
```
Comment (jgroth) [inline src/components/text-editor/text-editor.tsx:98]: Perhaps this should be marked as private as well? :slightly_smiling_face:

File: src/components/text-editor/types.ts (hunk 2 above)
Comment (civing) [reply src/components/text-editor/types.ts:8]: I think you might have conflated NodeConfig with NodeSpec? :)

File: src/components/text-editor/types.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+};
+
+export interface NodeConfig {
+    tagName: string;
+    attrs: string[];
+}
+
+export type NodeSpecFactory = (config: NodeConfig) => NodeSpec;
```
Comment (civing) [inline src/components/text-editor/types.ts:16]: I'm not sure we need to export this to consumers

File: src/components/text-editor/types.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * set to private to avoid usage while under development
+ */
+export type TextEditorPlugin = {
+    node: NodeConfig[];
+};
+
+export interface NodeConfig {
+    tagName: string;
+    attrs: string[];
+}
```
Comment (civing) [inline src/components/text-editor/types.ts:14]: Can we use the newly added `CustomComponent` used for the whitelist in the markdown reader? That was my intention at least when I created that one.

**Review by civing — COMMENTED** [review]
I'm yet to review all of it. But here are some initial comments and thoughts

File: src/components/text-editor/types.ts (hunk 2 above)
Comment (jgroth) [reply src/components/text-editor/types.ts:8]: Yes :slightly_smiling_face:

File: src/components/text-editor/text-editor.tsx (hunk 3 above)
Comment (john-traas) [reply src/components/text-editor/text-editor.tsx:103]: Yes :)

File: src/components/text-editor/types.ts (hunk 4 above)
Comment (john-traas) [reply src/components/text-editor/types.ts:16]: Good point. We can house it directly within the nodespec factory

File: src/components/text-editor/types.ts (hunk 5 above)
Comment (john-traas) [reply src/components/text-editor/types.ts:14]: Sounds good

**Review by civing — DISMISSED** [review]
Great!

**Review by civing — APPROVED** [review]

File: etc/lime-elements.api.md [hunk 6]
```diff
@@ -564,6 +564,8 @@ export namespace Components {
     export interface LimelProsemirrorAdapter {
         "contentType": 'markdown' | 'html';
         "language": Languages;
+        // Warning: (ae-incompatible-release-tags) The symbol "plugins" is marked as @beta, but its signature references "CustomElement" which is marked as @alpha
+        "plugins": CustomElement[];
```
Comment (adrianschmidt) [inline etc/lime-elements.api.md:568]: This warning means that we should mark the `plugins` prop as `@alpha`. I doesn't have an explicit `@beta` tag right now, but it, presumably, inherits the `@beta` tag from the class.

I'll see if adding `@alpha` to the prop removes this warning.

Comment (adrianschmidt) [reply etc/lime-elements.api.md:568]: #3263

**Pull request merged**